File: Stack/Leetcode/Hard/MaximalRectangle/MaximalRectangle.java
// https://leetcode.com/problems/maximal-rectangle/

import java.util.ArrayDeque;
import java.util.Deque;

public class MaximalRectangle {
    public int largestRectangleArea(int[] heights) {
        // find nse, pse
        // for each index to rectangle will be height[i] * (nse[i] - pse[i] - 1)
        Deque<Integer> st = new ArrayDeque<>();
        int maxR = 0;
        for(int i=0; i<heights.length; i++) {
            while(!st.isEmpty() && heights[st.peek()] >= heights[i]) {
                int ind = st.pop();
                int left = st.isEmpty() ? -1 : st.peek();
                maxR = Math.max(maxR, heights[ind] * (i - left - 1));
            }
            st.push(i);
        }
        while(!st.isEmpty()) {
            int ind = st.pop();
            int left = st.isEmpty() ? -1 : st.peek();
            maxR = Math.max(maxR, heights[ind] * (heights.length - left - 1));
        }
        return maxR;
    }
    public int maximalRectangle(char[][] matrix) {
        int[] pref = new int[matrix[0].length];
        for(int i=0; i<matrix[0].length; i++) {
            pref[i] = matrix[0][i] - '0';
        }
        int maxR = largestRectangleArea(pref);

        for(int i=1; i<matrix.length; i++) {
            for(int j=0; j<matrix[0].length; j++) {
                pref[j] = matrix[i][j] == '0' ? 0 : pref[j] + 1; 
            }
            maxR = Math.max(maxR, largestRectangleArea(pref));
        }
        return maxR;
    }
}

File: Stack/Leetcode/Hard/MaximalRectangle/MaximalRectangle.cpp
// https://leetcode.com/problems/maximal-rectangle/
#include<bits/stdc++.h>
using namespace std;
int largestRectangleArea(vector<int> &heights)
{
    stack<int> st;
    int n = heights.size();
    int maxArea = 0;
    for (int i = 0; i < n; i++)
    {
        while (!st.empty() && heights[st.top()] >= heights[i])
        {
            int ind = st.top();
            st.pop();
            int left = st.empty() ? -1 : st.top();
            maxArea = max(maxArea, heights[ind] * (i - left - 1));
        }
        st.push(i);
    }
    while (!st.empty())
    {
        int ind = st.top();
        st.pop();
        int left = st.empty() ? -1 : st.top();
        maxArea = max(maxArea, heights[ind] * (n - left - 1));
    }
    return maxArea;
}

public:
int maximalRectangle(vector<vector<char>> &matrix)
{
    vector<int> pref(matrix[0].size(), 0);
    for (int i = 0; i < matrix[0].size(); i++)
    {
        pref[i] = matrix[0][i] - '0';
    }
    int maxR = largestRectangleArea(pref);
    for (int i = 1; i < matrix.size(); i++)
    {
        for (int j = 0; j < matrix[0].size(); j++)
        {
            pref[j] = matrix[i][j] == '0' ? 0 : pref[j] + 1;
        }
        maxR = max(maxR, largestRectangleArea(pref));
    }
    return maxR;
}